### ISSUE TITLE
[Wayland Bug] latest version doesn't work properly on Wayland

### ISSUE THREAD

**Issue by ceyhunn** [paradoxxxzero/gnome-shell-system-monitor-applet#439]
It is Ok on Xorg. But with latest version ( version 34) some screen lagging, mouse lagging occurs. When I disable this extension, System works fine. More info: with previous version (version 32) it also works fine on Wayland.
I use Manjaro Gnome latest version.

**Comment by piegamesde** [comment]
Please have a look if this is related to #433 by not polling the processor's frequency or install the latest development version on my fork. (Just clone and copy the folder to `.local/share/gnome-shell/extensions`). I'll try Wayland the next time I log in.

**Comment by ceyhunnabiyev** [comment]
Hi @piegamesde , 
Yes. The same problem. I have installed development version. No success on Wayland.

**Comment by piegamesde** [comment]
@ceyhunnabiyev  Can you please disable all features (in the tray as well as in the menu) except one at the time to narrow down the issue further? I have no clue and can't reproduce it.

**Comment by pbasista** [comment]
I have experienced a similar issue, which was not specific to Wayland, but also occurred on Xorg version of Gnome. Its symptoms were, however, similar. The entire screen appeared to be freezing for a few milliseconds every ~1.5 seconds.

It turned out that the issue was caused by this extension and more specifically by the CPU frequency monitor. When I turned that part off, the issue has disappeared.

Perhaps the way how CPU frequency is obtained for the purpose of displaying it by this extension needs to be re-examined. Maybe there is some blocking part which also blocks the entire GNOME Shell UI.

**Comment by 14mRh4X0r** [comment]
I can confirm that this is caused by polling the frequency; Showing the graph or the text in the menu triggers it, with both off there's no freeze.

My guess is that reading `$ncpu` files is taking too long. The below test in bash on my system shows it takes about 135 ms.

```bash
nproc=$(nproc)
time for ((i = 0; i < nproc; i++)); do
  cat /sys/devices/system/cpu/cpu$i/cpufreq/scaling_cur_freq >/dev/null
done
```

**Comment by 14mRh4X0r** [comment]
I've tried to debug this issue, but when I change the for-loop in `Freq.refresh` to the following, I still get freezes:

```js
        let i = 0;
        let file = Gio.file_new_for_path(`/sys/devices/system/cpu/cpu${i}/cpufreq/scaling_cur_freq`);
        file.load_contents_async(null, Lang.bind(this, function cb (source, result) {
            let as_r = source.load_contents_finish(result);
            total_frequency += parseInt(as_r[1]);

            if (i + 1 >= num_cpus) {
                this.freq = Math.round(total_frequency / num_cpus / 1000);
            } else {
                file = Gio.file_new_for_path(`/sys/devices/system/cpu/cpu${i}/cpufreq/scaling_cur_freq`);
                file.load_contents_async(null, Lang.bind(this, cb));
            }
        });
```

This might just be me not understanding what `load_contents_async` does though, I just copied the usage from `Thermal.refresh`.

I also tested the access to those file a bit more. It seems like accessing the individual files costs 19 ms on my system, unless they're accessed more often. Accessing the file for cpu0 1000 times takes approximately 740 ms, while requesting them 25 times with a one second interval takes 25.497 seconds (for comparison, sleeping 1 second 25 times costs 25.060 seconds).

**Comment by 14mRh4X0r** [comment]
Sorry for the multicomment, but I just realized I forgot to restart gnome-shell while testing. The code I wrote above (with the edit to prevent an endless loop – oops) does fix the issue for me.

**Comment by return42** [comment]
@ceyhunnabiyev @14mRh4X0r : I made a fork [1] with the patch in @14mRh4X0r suggested. For me it fixed a lot of my issues I observed when system-monitor is activated.

Do you like to test my fork? .. any feedback is welcome / Thanks

[1] https://github.com/return42/gnome-shell-system-monitor-applet

**Comment by return42** [comment]
@14mRh4X0r  @piegamesde @pbasista This issue is fixed in master since PR #497 is merged. 
@ceyhunnabiyev  Can you please test current master and close this issue? / Thanks!
